File: algorithms/[90]子集 II/solution.py
# topics = ["数组", "哈希表"]

from typing import List, Set, Tuple


class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        """
        广度优先搜索 & 哈希表
        """
        nums.sort()

        # 集合 set 的元素只能是不可变值
        s: Set[Tuple[int, ...]] = set()
        s.add(tuple())

        for num in nums:
            # 复制一个相同的集合进行循环，避免 RuntimeError: Set changed size during iteration
            # 结果同 s.update([t + (num,) for t in s])
            for t in s.copy():
                s.add(t + (num,))

        return list(map(list, s))

File: algorithms/[90]子集 II/solution.ts
// topics = ["数组", "回溯算法"]

function subsetsWithDup(nums: number[]): number[][] {
  nums.sort((a, b) => a - b);
  const n = nums.length;
  const res: number[][] = [];
  const path: number[] = [];

  const backtrack = (idx = 0): void => {
    res.push([...path]);

    for (let i = idx; i < n; i += 1) {
      // 过滤掉重复子集
      if (i > idx && nums[i] == nums[i - 1]) continue;
      path.push(nums[i]);
      backtrack(i + 1);
      path.pop();
    }
  };

  backtrack();

  return res;
}
